### PR TITLE
Script for distilling zero-shot classifier to more efficient student

### PR DESCRIPTION
**Pull request by joeddav**
This PR introduces a script that provides a way to improve the speed and memory performance of a zero-shot classifier by training a more efficient student model from the zero-shot teacher's predictions over an unlabeled dataset.

For a given sequence, the zero-shot classification pipeline requires each possible label to be fed through the large NLI model separately. This requirement slows results considerably, particularly for tasks with a large number of classes `K`.

Given (1) an unlabeled corpus and (2) a set of candidate class names, this script allows a user to train a standard classification head with `K` output dimensions. The script generates a softmax distribution for the provided data & class names, and a student classifier is then fine-tuned on these proxy labels. The resulting student model can be used for classifying novel text instances over these `K` classes with an order-of-magnitude boost in inference speed in addition to decreased memory usage.

A teacher NLI model can be distilled to a student model by running `distill_classifier.py` like so:

```
python distill_classifier.py \
--data_file unlabeled_data.txt \
--class_names_file class_names.txt \
--output_dir ./distilled_model
```

A number of other args are provided as well, such as `--teacher_name_or_path` and `--student_name_or_path` for specifying the pre-trained student & teacher models to be used (by default `roberta-large-mnli` and `distillbert-base-uncased`) and `--hypothesis_template` for customizing the [hypothesis template](https://huggingface.co/transformers/main_classes/pipelines.html#transformers.ZeroShotClassificationPipeline.__call__) used by the teacher zero-shot model. The training is implemented via `Trainer`, so any [`TrainingArguments`](https://huggingface.co/transformers/main_classes/trainer.html#trainingarguments) can be specified as well.

The resulting model can then be used trivially in a text classification pipeline or in any other way:

```python
model = AutoModelForSequenceClassification.from_pretrained("./distilled_model")
tokenizer = AutoTokenizer.from_pretrained("./distilled_model")
distilled_classifier = TextClassificationPipeline(model=model, tokenizer=tokenizer)
```

See the included [README.md](https://github.com/joeddav/transformers/blob/zero-shot-distillation/examples/research_projects/zero-shot-distillation/README.md) for more details and examples.

Soon I'll introduce a similar script for self-training an NLI model, boosting the model's performance after training on only unlabeled data, which model can then be subsequently distilled with this script like any NLI model.

**Update**: I also just added a link to a working [colab notebook demo](https://colab.research.google.com/drive/1mjBjd0cR8G57ZpsnFCS3ngGyo5nCa9ya?usp=sharing).

### CODE REVIEW

File: examples/research_projects/zero-shot-distillation/README.md [hunk 1]
```diff
@@ -0,0 +1,144 @@
+# Zero-shot classifier distillation
+
+This script provides a way to improve the speed and memory performance of a zero-shot classifier by training a more
+efficient student model from the zero-shot teacher's predictions over an unlabeled dataset.
```
Comment (LysandreJik) [inline examples/research_projects/zero-shot-distillation/README.md:4]: Could you add your github handle here?

```suggestion
# Zero-shot classifier distillation

Author: @joeddav 

This script provides a way to improve the speed and memory performance of a zero-shot classifier by training a more
efficient student model from the zero-shot teacher's predictions over an unlabeled dataset.
```

**Review by LysandreJik — APPROVED** [review]
Fantastic that you're using the `Trainer` for that. Pinging Sylvain for review.

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+    if training_args.do_eval:
+        agreement = trainer.evaluate(eval_dataset=dataset)["eval_agreement"]
+        logger.info(f"Agreement of student and teacher predictions: {agreement * 100:0.2f}%")
+
+    trainer.save_model()
+
+
+if __name__ == "__main__":
+    main()
```
Comment (LysandreJik) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:320]: TPU support could probably be added very easily by adding the following

```suggestion
def _mp_fn(index):
    # For xla_spawn (TPUs)
    main()
    
if __name__ == "__main__":
    main()
```

Comment (joeddav) [reply examples/research_projects/zero-shot-distillation/distill_classifier.py:320]: This would work for the training step, but generating the pseudo-labels isn't done with Trainer so I'd have to add TPU support manually for that piece. I can do that if TPU inference is straightforward (I've yet to try it out), but otherwise I'd probably rather get this out there now and add support for other things like that in a separate PR.

File: examples/research_projects/zero-shot-distillation/README.md [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+### Usage
+
+A teacher NLI model can be distilled to a more efficient student model by running `distill_classifier.py`:
+
+```
+python distill_classifier.py \
+--data_file <unlabeled_data.txt> \
+--class_names_file <class_names.txt> \
+--output_dir <output_dir>
+```
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/README.md:29]: Since it uses `Trainer`, it works out of the box for distributed training. Could you add an example of command launching the script on two GPUs? Users often ask us for those.

File: examples/research_projects/zero-shot-distillation/README.md [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+```python
+class_names = ["the world", "sports", "business", "science/tech"]
+hypothesis_template = "This text is about {}."
+sequence = "A new moon has been discovered in Jupiter's orbit"
+
+zero_shot_classifier = pipeline("zero-shot-classification", model="roberta-large-mnli")
+zero_shot_classifier(sequence, class_names, hypothesis_template=hypothesis_template)
+# {'sequence': "A new moon has been discovered in Jupiter's orbit",
+#  'labels': ['science/tech', 'the world', 'business', 'sports'],
+#  'scores': [0.7035840153694153, 0.18744826316833496, 0.06027870625257492, 0.04868902638554573]}
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/README.md:77]: Use the doctest syntax here:
```suggestion
>>> class_names = ["the world", "sports", "business", "science/tech"]
>>> hypothesis_template = "This text is about {}."
>>> sequence = "A new moon has been discovered in Jupiter's orbit"

>>> zero_shot_classifier = pipeline("zero-shot-classification", model="roberta-large-mnli")
>>> zero_shot_classifier(sequence, class_names, hypothesis_template=hypothesis_template)
{'sequence': "A new moon has been discovered in Jupiter's orbit",
 'labels': ['science/tech', 'the world', 'business', 'sports'],
 'scores': [0.7035840153694153, 0.18744826316833496, 0.06027870625257492, 0.04868902638554573]}
```

File: examples/research_projects/zero-shot-distillation/README.md [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+and even used trivially with a `TextClassificationPipeline`:
+
+```python
+distilled_classifier = TextClassificationPipeline(model=model, tokenizer=tokenizer, return_all_scores=True)
+distilled_classifier(sequence)
+# [[{'label': 'the world', 'score': 0.14899294078350067},
+#   {'label': 'sports', 'score': 0.03205857425928116},
+#   {'label': 'business', 'score': 0.05943061783909798},
+#   {'label': 'science/tech', 'score': 0.7595179080963135}]]
+```
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/README.md:118]: Same here

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+    HfArgumentParser,
+    Trainer,
+    TrainingArguments,
+    set_seed,
+    utils,
+)
+from transformers.trainer_utils import get_last_checkpoint, is_main_process
+
+
+DESCRIPTION = """\
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:24]:
```suggestion
DESCRIPTION = """
```

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    do_train: bool = field(default=True, metadata={"help": "Whether to run training of student model."})
+    do_eval: bool = field(
+        default=True,
+        metadata={
+            "help": (
+                "Whether to evaluate the agreement of the final student predictions and the teacher predictions"
+                "after training."
+            )
+        },
+    )
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:105]: All those arguments are already in `TrainingArguments`, so this class is the same as `TrainingArguments`. Not sure what you were trying to achieve.

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        target_p = inputs["labels"]
+        outputs = model(inputs["input_ids"], attention_mask=inputs["attention_mask"])
+        logits = outputs[0]
+
+        loss = -torch.sum(target_p * logits.log_softmax(dim=-1), axis=-1).mean()
+
+        if return_outputs:
+            return loss, outputs
+
+        return loss
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:119]: Great example of Trainer subclass!

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+            batch_premises,
+            batch_hypotheses,
+            padding=True,
+            truncation="only_first",
+            return_tensors="pt",
+        )
+
+        with torch.no_grad():
+            outputs = model(**encodings)
+            logits.append(outputs.logits.detach().cpu())
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:188]: I think we could also pass along the fp16 flag and wrap this in an `autocast` here to speed up predictions?

File: examples/research_projects/zero-shot-distillation/distill_classifier.py (hunk 2 above)
Comment (sgugger) [reply examples/research_projects/zero-shot-distillation/distill_classifier.py:320]: Looking at the rest of the script, it would work. Just each TPU would generate all the pseudo-labels. But that's also the case for multi-GPU at the moment.

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+    no_cuda: bool,
+):
+    """
+    Gets predictions by the same method as the zero-shot pipeline but with DataParallel & more efficient batching
+    """
+    model = AutoModelForSequenceClassification.from_pretrained(model_path)
+    model_config = model.config
+    if not no_cuda and torch.cuda.is_available():
+        model = nn.DataParallel(model)
+        batch_size *= len(model.device_ids)
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:168]: This should be more restrictive: if launched with torch.distributed.launch, I fear this part will try to use all the GPUs number of GPUs times (once for each process launched).

**Review by sgugger — APPROVED** [review]
Great new example! I'm not sure it does support distributed training after reading everything so the PR should either be amended to support it or clearly indicate in the README it does not support it (same for TPUs).

File: examples/research_projects/zero-shot-distillation/distill_classifier.py (hunk 7 above)
Comment (joeddav) [reply examples/research_projects/zero-shot-distillation/distill_classifier.py:105]: It subclasses `TrainingArguments` and overrides the default values.

**Comment by joeddav** [comment]
@sgugger cool thanks for the feedback. Added `fp16` for the teacher predictions. It will now throw an error if someone tries to run it w/ distributed or TPUs. It _can_ do multi-gpu though and will do so automatically if multiple GPUs are available on the machine, it just can't do multi-node.

**Comment by sgugger** [comment]
Yes I meant distributed multi-GPU. I did see it will use all GPUs available on the machine however :-)

File: examples/research_projects/zero-shot-distillation/distill_classifier.py [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+    # Set the verbosity to info of the Transformers logger (on main process only):
+    if is_main_process(training_args.local_rank):
+        utils.logging.set_verbosity_info()
+        utils.logging.enable_default_handler()
+        utils.logging.enable_explicit_format()
+
+    if training_args.local_rank != -1:
+        raise ValueError("Distributed training is not currently supported.")
+    if training_args.tpu_num_cores is not None:
+        raise ValueError("TPU acceleration is not currently supported.")
```
Comment (sgugger) [inline examples/research_projects/zero-shot-distillation/distill_classifier.py:259]: Great!

**Review by sgugger — APPROVED** [review]
